graphql_compiler/compiler/ir_lowering_common/common.py

```diff
@@ -173,15 +181,22 @@ def visitor_fn(expression):
         else:
             return expression
 
-        expression_to_rewrite = None
+        expression_to_rewrite: Optional[BinaryComposition] = None
+        # N.B. the type checks in the last 2 elif statements are disabled because of the following:
+        # ```
+        # Incompatible types in assignment (expression has type "Expression", variable has type
+        # "Optional[BinaryComposition]")
+        # ```
+        # However, given the logic, we know that `right_is_binary_composition` is even more
+        # specifically that the right is BinaryComposition, not a basic Expression
         if expression.left == identity_literal and right_is_binary_composition:
             return expression.right
         elif expression.right == identity_literal and left_is_binary_composition:
             return expression.left
         elif expression.left == inverse_literal and right_is_binary_composition:
-            expression_to_rewrite = expression.right
+            expression_to_rewrite = expression.right  # type: ignore
         elif expression.right == inverse_literal and left_is_binary_composition:
-            expression_to_rewrite = expression.left
+            expression_to_rewrite = expression.left  # type: ignore
```

Comment:
The way this is written made typing a tad weird here. I would consider re-writing some fo the logic here, but I think it is outside the scope of this diff.

Reply:
Would a `cast` be appropriate instead of `type: ignore`? `type: ignore` prevents a good chunk of type checking onward from this point, whereas `cast` suppresses this typing issue but preserves the stated new type, and type checking may resume onward from here.

Reply:
Brilliant, didn't know about this.